### PR TITLE
feat: :sparkles: disable/enable LoRas with a switch

### PR DESCRIPTION
**Pull request by joshistoast**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [ ] Yes
- [x] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [x] No

## Description
Added an easy way to enable/disable LoRas within the ui. This is done by adding a switch next to the delete button, with logic that sets the weight to 0 when unchecked (disabled), and restoring the LoRa weight to it's last value when checked again (re-enabled).

## Related Tickets & Documents

- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings
image

## Added/updated tests?

- [ ] Yes
- [x] No : I don't wanna

### CODE REVIEW

**Comment by joshistoast** [comment]
Increasing/decreasing the LoRa count badge would be neat, might change that later

**Review by psychedelicious — CHANGES_REQUESTED** [review]
Thanks Josh. This has a few issues:
1. If you disable a lora and refresh the page, the user-selected lora weight will be reset to zero, losing their setting
2. Due to the issue 1, the switch won't let you enable the lora again - the value in state is zero
3. The "disabled" loras will still be added to the graph (this is a pre-existing issue, loras with weight of 0 are added to the graph, it's just more of a problem when the concept of "disabling" is introduced to loras)

A better way to handle this would be to add `isEnabled` to the lora redux state.

**Comment by joshistoast** [comment]
> Thanks Josh. This has a few issues:
> 
>     1. If you disable a lora and refresh the page, the user-selected lora weight will be reset to zero, losing their setting
> 
>     2. Due to the issue 1, the switch won't let you enable the lora again - the value in state is zero
> 
>     3. The "disabled" loras will still be added to the graph (this is a pre-existing issue, loras with weight of 0 are added to the graph, it's just more of a problem when the concept of "disabling" is introduced to loras)
> 
> 
> A better way to handle this would be to add `isEnabled` to the lora redux state.

Understood, I can use redux state to track that. However I feel like the first 2 can easily be fixed, and the existence of the lora within the graph shouldn't affect the generation if it's weight is 0 (unless for some reason it does).

**Comment by hipsterusername** [comment]
> > Thanks Josh. This has a few issues:
> > ```
> > 1. If you disable a lora and refresh the page, the user-selected lora weight will be reset to zero, losing their setting
> > 
> > 2. Due to the issue 1, the switch won't let you enable the lora again - the value in state is zero
> > 
> > 3. The "disabled" loras will still be added to the graph (this is a pre-existing issue, loras with weight of 0 are added to the graph, it's just more of a problem when the concept of "disabling" is introduced to loras)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > A better way to handle this would be to add `isEnabled` to the lora redux state.
> 
> Understood, I can use redux state to track that. However I feel like the first 2 can easily be fixed, and the existence of the lora within the graph shouldn't affect the generation if it's weight is 0 (unless for some reason it does).

Think it’s generally going to be bad practice to leave nodes in a composed graph that shouldn’t be there. It creates potential for future bugs/issues to crop up with graph mutation

**Comment by joshistoast** [comment]
@psychedelicious Where would I exclude disabled LoRas from the graph, given they now have a `disabled` flag that can be used?

File: invokeai/frontend/web/src/features/lora/store/loraSlice.ts [hunk 1]
```diff
@@ -58,16 +59,35 @@ export const loraSlice = createSlice({
       }
       lora.weight = defaultLoRAConfig.weight;
     },
+    loraToggle: (state, action: PayloadAction<{ id: string; disabled: boolean }>) => {
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/lora/store/loraSlice.ts:62]: Another naming nitpick - it should be `loraToggled`, or `loraIsEnabledChanged`.

File: invokeai/frontend/web/src/features/settingsAccordions/components/GenerationSettingsAccordion/GenerationSettingsAccordion.tsx [hunk 2]
```diff
@@ -31,7 +31,7 @@ const formLabelProps: FormLabelProps = {
   minW: '4rem',
 };
 
-const badgesSelector = createMemoizedSelector(selectLoraSlice, selectGenerationSlice, (lora, generation) => {
+const badgesSelector = createMemoizedSelector(selectNonDisabledLoraSlice, selectGenerationSlice, (lora, generation) => {
   const loraTabBadges = size(lora.loras) ? [size(lora.loras)] : [];
   const accordionBadges: (string | number)[] = [];
   if (generation.model) {
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/settingsAccordions/components/GenerationSettingsAccordion/GenerationSettingsAccordion.tsx:37]: I'd prefer to not use an extra selector here - all we need is the count of enabled loras for the badge. So something like this is sufficient:

```ts
const enabledLoRAsCount = filter(lora.loras, (l) => l.isEnabled).length;
const loraTabBadges = enabledLoRAsCount ? [enabledLoRAsCount] : [];
```

File: invokeai/frontend/web/src/features/lora/components/LoRACard.tsx
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/lora/components/LoRACard.tsx:1]: Let's disable the weight slider and number input when the lora is disabled

File: invokeai/frontend/web/src/features/nodes/util/graph/addLoRAsToGraph.ts [hunk 3]
```diff
@@ -28,8 +28,9 @@ export const addLoRAsToGraph = (
    * So we need to inject a LoRA chain into the graph.
    */
 
-  const { loras } = state.lora;
-  const loraCount = size(loras);
+  // const { loras } = state.lora;
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/nodes/util/graph/addLoRAsToGraph.ts:31]: errant comment

File: invokeai/frontend/web/src/features/settingsAccordions/components/GenerationSettingsAccordion/GenerationSettingsAccordion.tsx [hunk 4]
```diff
@@ -32,7 +33,8 @@ const formLabelProps: FormLabelProps = {
 };
 
 const badgesSelector = createMemoizedSelector(selectLoraSlice, selectGenerationSlice, (lora, generation) => {
-  const loraTabBadges = size(lora.loras) ? [size(lora.loras)] : [];
+  const enabledLoRAsCount = filter(lora.loras, (l: LoRA) => !!l.isEnabled).length;
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/settingsAccordions/components/GenerationSettingsAccordion/GenerationSettingsAccordion.tsx:36]: Doesn't need the `LoRA` type annotation here

File: invokeai/frontend/web/src/features/lora/store/loraSlice.ts [hunk 5]
```diff
@@ -58,10 +60,18 @@ export const loraSlice = createSlice({
...
       lora.weight = defaultLoRAConfig.weight;
     },
+    loraToggled: (state, action: PayloadAction<Pick<LoRA, 'id' | 'isEnabled'>>) => {
+      const { id, isEnabled } = action.payload;
+      const lora = state.loras[id];
+      if (!lora) {
+        return;
+      }
+      lora.isEnabled = isEnabled;
+    },
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/lora/store/loraSlice.ts:70]: "toggled" implies the new state will be the opposite of the previous state - so I'd expect "toggled" actions to not have an `isEnabled` boolean. `loraIsEnabledChanged` is clearer.

This is the establish project convention, used elsewhere, for ex:

```ts
    controlAdapterIsEnabledChanged: (state, action: PayloadAction<{ id: string; isEnabled: boolean }>) => {
    ...
    controlAdapterAutoConfigToggled: (state, action: PayloadAction<{ id: string }>) => {
```

**Review by psychedelicious — CHANGES_REQUESTED** [review]
Just a couple minor changes requested, ty

**Review by psychedelicious — APPROVED** [review]
Thanks @joshistoast ! Nice UX win here.

**Pull request merged**